File: solutions/868.push-dominoes/push-dominoes.cpp
class Solution {
public:
    string pushDominoes(string dominoes) {
        int len = dominoes.length();
        vector<int> visit(len);
        for(int i = 0; i < len; i++) {
            for(int j = 0; j < len; j++) {
                if(dominoes[j] == 'R' && j < len && dominoes[j + 1] == '.') 
                    visit[j + 1] += 1;
                if(dominoes[j] == 'L' && j > 0 &&dominoes[j - 1] == '.') 
                    visit[j - 1] -= 1;
            }
            int flag = 0;
            for(int i = 0; i < len; i++) {
                if(dominoes[i] == '.' && visit[i] != 0) {
                    if(visit[i] == 1) dominoes[i] = 'R';
                    else dominoes[i] = 'L';
                    flag = 1;
                }
            }
            if(flag == 0) return dominoes;
            for(int i = 0; i < len; i++) visit[i] = 0;
        }
        return dominoes;
    }
};

File: solutions/868.push-dominoes/push-dominoes.java
class Solution {
    public String pushDominoes(String dominoes) {
        char[] dominArr = dominoes.toCharArray();
        for (int i = 0; i < dominArr.length; i++) {
            if (dominArr[i] == 'L') {
                for (int j = i - 1; j >= 0; j--) {
                    if (dominArr[j] == '.') {
                        dominArr[j] = 'L';
                    } else {
                        break;
                    }
                }
            } else if (dominArr[i] == 'R') {
                int Lindex = dominoes.indexOf('L', i+1);
                if (Lindex == -1) {
                    for (int j = i + 1; j < dominArr.length; j++) 
                        dominArr[j] = 'R';
                    break;
                } else {
                    int Rindex=0, j = Lindex - 1;
                    while(true) {
                        if (dominArr[j] == 'R') {
                            Rindex = j;
                            break;
                        }
                        j--;
                    }
                    while(i < Rindex) dominArr[i++] = 'R';
                    j = Lindex;
                    while (i < j) {
                        dominArr[i++] = 'R';
                        dominArr[j--] = 'L';
                    }
                    i = Lindex;
                }
            }
        }
        return new String(dominArr);
    }
}